playwright-test: Add PLAYWRIGHT_BROWSERS_PATH to build environment

the-sun-will-rise-tomorrow opened the pull request:
This allows writing self-contained node scripts (using a `nix-shell` shebang), without requiring a separate `shell.nix` which sets `PLAYWRIGHT_BROWSERS_PATH` to `pkgs.playwright-driver.browsers`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

the-sun-will-rise-tomorrow committed:
playwright-test: Add PLAYWRIGHT_BROWSERS_PATH to build environment

This allows writing self-contained node scripts (using a nix-shell
shebang), without requiring a separate shell.nix which sets
PLAYWRIGHT_BROWSERS_PATH to pkgs.playwright-driver.browsers.

kalekseev commented on pkgs/development/web/playwright/driver.nix:
```diff
@@ -191,9 +192,25 @@ let
 
+    setupHook = writeText "setupHook.sh" ''
+      addBrowsersPath () {
+        export PLAYWRIGHT_BROWSERS_PATH="${playwright-core.passthru.browsers}"
```
Is it possible to override this var? if user wants to run his script with different browsers path

the-sun-will-rise-tomorrow replied:
Sure, you can override it as usual (by setting it in the environment before running the script).

I think overlays etc. should still work too? Or do we need to use `pkgs.playwright-driver` instead of the local `playwright-core` definition to pull in customizations? There's a few references to `playwright-core` already...

kalekseev replied:
Playwright package named `core` but in nixpkgs it was named `driver` initially and we keep it for backward compatibility. Would be nice to migrate to `core` eventually but at the moment playwright PRs are waiting for months before someone review and merge them so let's use `driver`

the-sun-will-rise-tomorrow replied:
Sorry, I don't mean the name, but whether we're using the definition from `pkgs` so that overrides apply. I think that if we want to make packages overridable with things like overlays or `overrideAttrs`, then we have to use `finalAttrs` or import ourselves and use that to refer to our package instead of using what we defined directly.

Let me try a few things...

the-sun-will-rise-tomorrow replied:
Actually I don't see any way to override the browser derivation.

I see that in https://github.com/NixOS/nixpkgs/pull/344614 the browsers were wrapped in `makeOverridable`, but that was so that a user can customize and use just the browsers (by setting `PLAYWRIGHT_BROWSERS_PATH` to point to them), not for building a Node or Python package that uses a different set of browsers in its wrapped scripts.

So, it looks like doing it that way was never possible, but we are not making it worse here.

the-sun-will-rise-tomorrow replied:
> Is it possible to override this var? if user wants to run his script with different browsers path

Actually forgot about the case if the user sets the variable before running `nix-shell` (and doesn't use `--pure`). Changed it so we set it only if it's unset.

kalekseev reviewed (commented):
Do we still need  ` --set-default PLAYWRIGHT_BROWSERS_PATH "${playwright-core.passthru.browsers}"` in makeWrapper?

Also the python library needs the same changes.

the-sun-will-rise-tomorrow committed:
python3Packages.playwright: Add PLAYWRIGHT_BROWSERS_PATH to build environment

This allows writing self-contained Python scripts (using a nix-shell
shebang), without requiring a separate shell.nix which sets
PLAYWRIGHT_BROWSERS_PATH to pkgs.playwright-driver.browsers.

the-sun-will-rise-tomorrow commented:
> Do we still need ` --set-default PLAYWRIGHT_BROWSERS_PATH "${playwright-core.passthru.browsers}"` in makeWrapper?

Yes, it's still useful, as that creates a `praywright` command that will run with the correct value from any environment. The `setupHook` can only apply to things that use the package as a dependency (which includes nix-shell and nix-shell shebangs).

> Also the python library needs the same changes.

OK, added.

kalekseev commented on pkgs/development/python-modules/playwright/default.nix:
```diff
@@ -101,6 +112,9 @@ buildPythonPackage rec {
       {
         driver = playwright-driver;
         browsers = playwright-driver.browsers;
+        env = runCommand "playwright-env-test" {
```
we already have playwright test that uses python package, I guess we can remove this var in it https://github.com/nixos/nixpkgs/blob/26713f779299cd4688ef2f35e7f19cbc2e207f79/nixos/tests/playwright-python.nix#L15 to test the setupHook

the-sun-will-rise-tomorrow replied:
I think this change will not apply to that test. There, we don't run Playwright from a derivation directly, but instead use `writePython3Bin` which will create a command that we will then run from outside later, with a different environment that won't have that variable.

kalekseev reviewed (approved):
LGTM

K900 commented:
This now leaks Chromium into the closure of literally every single thing that depends on Playwright, which is bad.

K900 commented:
Should we just revert this?

kalekseev commented:
> Should we just revert this?

I'm not sure why something would depend on playwright without browsers (any examples?) but I guess people may want to choose only chromium browser using `playwright-driver.browsers-chromium` instead of `playwright-driver.browsers`.
So yeah we should revert and think about another solution something like having separate package `playwright-browserless`.

K900 commented:
Things like e.g. urlwatch have playwright as an optional dependency, where the code path isn't hit unless you use a specific configuration (which also specifies the browser to use). With this change, the default config will pull in an entire Chromium, even if it's entirely unused.
